File: solutions/lab2/ruby/live_stream_connection.rb
require 'net/http'
require 'logger'
require 'json'

module Lab2
  class LiveStreamConnection

    @@logger = Logger.new(STDOUT)

    def logger
      @@logger
    end

    def initialize(endpoint_url, access_token)
      @endpoint_url = endpoint_url
      @access_token = access_token
    end

    def open
      logger.debug "opening live stream connection to #{@endpoint_url}..."

      uri = URI(@endpoint_url)

      req = Net::HTTP::Get.new(uri)
      req['Authorization'] = "Bearer #{@access_token}"

      buffer = ""

      Net::HTTP.start(uri.hostname, uri.port, use_ssl: true) do |http|
	http.request req do |response|
	  response.read_body do |chunk|
	    # the streaming response is chunked transfer encoding, so we
	    # need to read chunks and emit records separated by newlines.
	    # note that records may span chunks...so we need to read the
	    # chunks into buffers and emit records when we see newlines.
            chunk.each_char do |c|
	      if c == "\n" then
		puts buffer
		calculate_lag(buffer)
		puts "-------------------------------------"
		buffer = ""
	      else
		buffer << c
	      end
	    end
	  end
	end
      end
    end

    def calculate_lag(record)
      record_json = JSON.parse(record)
      timestamp = record_json['receivedTimeGMT']
      if timestamp.nil? then
	looger.error "unable to find receivedTimeGMT field in record for lag calculation"
      else
	now = Time.new.to_i
	logger.info "calculate_lag: record lag is #{now - timestamp.to_i} seconds"
      end
    end

  end
end

File: lab2/java/src/lab2/LiveStreamConnection.java
package lab2;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.net.URL;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.BufferedReader;
import java.io.IOException;
import java.net.HttpURLConnection;
import javax.net.ssl.HttpsURLConnection;
import java.util.zip.GZIPInputStream;

import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.Map;

import lab.Credentials;

public class LiveStreamConnection {

  public static final Logger log = LoggerFactory.getLogger(LiveStreamConnection.class);

  protected String endpointUrl;
  protected String accessToken;

  public LiveStreamConnection(String endpointUrl, String accessToken)
  {
    // perform your instance initialization here
  }

  public void open() throws IOException {
    // insert your code here to connect to the live stream endpoint.
    // you will need to pass the access_token in a HTTP authorization header
    // that looks like this:
    //
    // Authorization: Bearer <hex access token string>
    // 
    // the streaming response is chunked transfer encoding, so we
    // need to read chunks and emit records separated by newlines.
    // note that records may span chunks...so we need to read the
    // chunks into buffers and emit records when we see newlines.
  }

  public void calculate_lag(String record) throws IOException {
    // you probably want a function like this to take a hit record in string format,
    // parse it into a JSON hash, extract the event timestamp from the receivedTimeGMT
    // field and calculate the difference between that timestamp and now.
  }
}
